cupyx/jit/_builtin_funcs.py

```diff
@@ -219,6 +219,34 @@ def call_const(self, env, dtype, size, alignment=None):
         env.locals[name] = var
         return Data(name, _cuda_types.Ptr(ctype))
 
+class LocalMemory(BuiltinFunc):
+
+    def __call__(self, dtype, size, alignment=None):
+        """Allocates local memory and returns it as a 1-D array.
+
+        Args:
+            dtype (dtype):
+                The dtype of the returned array.
+            size (int or tuple):
+                If ``int`` type, the size of static local memory.
+                If ``tuple`` type, the size of a multi-dimensional static local memory.
+                Does not use __extern__ keyword like shared memory does
+            alignment (int or None): Enforce the alignment via __align__(N).
+        """
+        super().__call__()
+
+    def call_const(self, env, dtype, size, alignment=None):
+        name = env.get_fresh_variable_name(prefix='_lmem')
+        ctype = _cuda_typerules.to_ctype(dtype)
+        var = Data(name, _cuda_types.LocalMem(ctype, size, alignment))
+        env.decls[name] = var
+        env.locals[name] = var
+        if(type(size) is int):
+            size = (size,)
+        return_ctype = ctype
+        for i in range(len(size)):
+            return_ctype = _cuda_types.ContiguousArray(return_ctype, size[len(size)-(i+1):])
+        return Data("!!lmem", return_ctype) #non-valid name in order to intercept creation of this
```

Comment:
The following code exposes CUDA error.
```py
@jit.rawkernel()
def f():
    lmem = jit.local_memory(numpy.int32, (2, 3, 4)) + 10
```

```
cubin.cu(7): error: a value of type "int" cannot be used to initialize an entity of type "int (*)[3][4]"
    int (*lmem)[3][4] = (!!lmem + 10);
```

Reply:
Hey, I recognize for this case it is non-ideal because it exposes the !!lmem token designed to catch the creation of local_memory, but to be honest I wouldn't have expected this code to work in the first place. Testing a similar expression in numba doesn't work either:

```
import numba
from numba import cuda, int32

@cuda.jit
def test():
    a = cuda.local.array((2,3,4), int32) + 10
    
test[256, 256]()
```
```
No implementation of function Function(<built-in function add>) found for signature:
 
 >>> add(array(int32, 3d, C), Literal[int](10))
```

Would it be desired to catch this particular case? I could imagine it could be a syntactic way to initialize the array, by generating the following for the above code:
`int lmem[2][3][4] = {10};`

However, this would almost certainly involve modifying something in cupyx.jit._compile to catch this edge case

Thanks again for all your help with this by the way! :)

Reply:
In case of shared memory, it generates a code that adds offset to the pointer.

```py
@jit.rawkernel()
def f():
    smem = jit.shared_memory(numpy.int32, 10) + 1
```

```cpp
extern "C" __global__ void f() {
   __shared__ int _smem1[10];
  int* smem = (_smem1 + 1);
}
```

Reply:
Interesting. If we wrap shared/local memory in an array-like fashion, I'd expect users to index over the arrays, instead of computing the pointer offset this way, which is only kinda intuitive for 1D arrays. Semantics-wise, adding an array with a constant should result in a broadcasting, i.e.
```py
smem = jit.local_memory(numpy.int32, (2,3)) + 1
```
is equivalent to
```c
int smem[2][3];
for (auto i=0; i<2; i++)
  for (auto j=0; j<3; j++)
    smem[i][j] += 1
```

Reply:
This can be done in one line as `int smem[2][3] = {1};` I believe, the brace notation with a single value inside should broadcast to the entire array when its created. (Also has the advantage of ensuring the array is indeed guaranteed to be initialized to some value, I've run into an issue in the past where a local array created using `int lmem[10]` had spurious non-zero values sprinkled throughout!)

Reply:
You missed the plus sign 🙂 Here, it's not just about initializing the array, it's about all possible, broadcastable arithmetic operations that NumPy/CuPy would permit. Interpreting the math operators as just calculating pointer offsets kills this possibility, and I want to make sure this is a fully-aware decision.

Reply:
Done a fair amount of thinking, and ultimately I agree that the syntax should follow python convention rather than pointer offsets. The more pythonic the syntax, the more people who will be able to just pick it up and use it. And cupy is python-based anyways...

However in this case, would it not make more sense to implement the following *as* local_memory to maximize the overlap in code?

```
@jit.rawkernel
def test():
    smem = cupy.zeros([2, 3])+1
```
(e.g.: if cupy.zeros is called by the CPU, it creates a device_array, if it's called by the GPU within a @jit.rawkernel decorated function, it creates a local_memory array)

It could lead to some confusion, but this seems like a natural extension

Reply:
For reference, my main use case for cupy is to create a "parallel for" loop of the form:

```
@jit.rawkernel
def kernel(array):
    tid = jit.grid(1)
    ntid = jit.gridsize(1)
    for i in range(tid, array.shape[0], ntid):
        #do stuff at each index in the array
```

Essentially an elementwise kernel, but with an index so each thread can potentially access neighboring data (partial differential equation simulations). 

However, this use case is relatively independent of the choice of going with pythonic syntax, or sticking with c, so as long as a local memory array exists in some capacity, I don't have much skin in the game :)

Reply:
Also just learned I was wrong, the syntax `int lmem[2][2] = {1};` apparently will only initialize the first element of the top-level array to one, creating `{{1, 1}, {0, 0}}`!

Reply:
> Also just learned I was wrong

Yeah, C++ is hard and full of surprises.

> However in this case, would it not make more sense to implement the following _as_ local_memory to maximize the overlap in code?
> 
> ```
> @jit.rawkernel
> def test():
>     smem = cupy.zeros([2, 3])+1
> ```
> 
> (e.g.: if cupy.zeros is called by the CPU, it creates a device_array, if it's called by the GPU within a @jit.rawkernel decorated function, it creates a local_memory array)
> 
> It could lead to some confusion, but this seems like a natural extension
> 
> (And for shared memory, potentially a flag could be added? `shared_mem = cupy.zeros([2,3], shared_memory=True)`)
> 
> This idea could be extended to all cupy functions: if called by the CPU, it runs a batched algorithm on a device_array. if called by the GPU, it runs the algorithm as if it was the corresponding numpy algorithm, but within a GPU kernel.

I think this is way out of scope for CuPy JIT, let's not overthink here 😅 I agree for your use case as long as `local_memory`/`shared_memory` are present, you're unblocked. The question I raised was mainly about their semantics (are they pythonic arrays, or should they be treated as pointers in disguise?), I didn't mean bringing any of the current host capabilities to `jit.rawkernel`.

Reply:
We don't want to allow element-wise operations in this notation in CuPy JIT, because it behaves defferently from CuPy ndarray routines in terms of parallelism.

Reply:
> because it behaves defferently from CuPy ndarray routines in terms of parallelism.

For shared memory, this syntax ambiguity is indeed a concern. For local memory, it's more like a syntactic sugar (see https://github.com/cupy/cupy/pull/7976#discussion_r1399454937). But it's better that we stick to the lowest common denominator and do not consider it.

Reply:
To be completely honest, I think the path of least resistance for this one is to tell users "don't do that". 

For the pointer offset, regardless of how you initialize the problem, you end up with two registers, one for the original array, and one for the offset array. May as well just initialize it as:

```
original_pointer = jit.local_memory(numpy.int32, 10)
offset_pointer = original_pointer + 1
```

For sending the array to a function upon creation, you wouldn't even have a variable handle for the array you've created in this case. Would it not exclusively be better to write it as the following (where you would have a variable to manipulate the array afterwards?):

```
@jit.rawkernel(device=True)
def g(array):
    pass

@jit.rawkernel()
def f():
    array = jit.local_memory(numpy.int32, 10)
    g(array)

f[1, 1]()
```

Reply:
Was mainly just wondering if that would be alright :) I'll get right on it!

Reply:
Oh! In that case this whole business with trying to catch the creation of the local memory array is moot. I'll rework it back to match the syntax of shared memory, in which case it should be able to do pointer offsets and submit to a function as-is like shared memory can.